lib/oban/engines/basic.ex

```diff
@@ -324,20 +324,43 @@ defmodule Oban.Engines.Basic do
     end
   end
 
-  defp return_or_replace(conf, query_opts, job, changeset) do
+  defp return_or_replace(conf, query_opts, %{state: state} = job, changeset) do
     case Changeset.get_change(changeset, :replace, []) do
       [] ->
         {:ok, job}
 
-      replace_keys ->
-        Repo.update(
-          conf,
-          Ecto.Changeset.change(job, Map.take(changeset.changes, replace_keys)),
-          query_opts
-        )
+      replace ->
+        if valid_replace_option?(replace) do
+          job_state = String.to_existing_atom(state)
+          replace_keys = Keyword.get(replace, job_state, [])
+
+          maybe_replace(conf, query_opts, job, changeset, replace_keys)
+        else
+          replace_keys(conf, query_opts, job, changeset, replace)
+        end
     end
   end
 
+  def valid_replace_option?(replace_keys) do
+    all_states = Oban.Job.states()
+
+    match?([{_, _} | _], replace_keys) and
+      replace_keys |> Keyword.keys() |> Enum.all?(&(&1 in all_states))
+  end
+
+  defp maybe_replace(_conf, _query_opts, job, _changeset, []), do: {:ok, job}
+
+  defp maybe_replace(conf, query_opts, job, changeset, replace_keys),
+    do: replace_keys(conf, query_opts, job, changeset, replace_keys)
+
+  defp replace_keys(conf, query_opts, job, changeset, replace_keys) do
```

Comment:
Let's avoid the `maybe` variant and rely on a function head:

```suggestion
  defp replace_keys(_conf, _query_opts, job, _changeset, []), do: {:ok, job}
  
  defp replace_keys(conf, query_opts, job, changeset, replace_keys) do
```